README.md
# UKCP18 CVs
Controlled Vocabularies (CVs) for use in UKCP18

## Using these vocabularies

We strongly recommend that you use the following recipe for working with these vocabularies:

### 1. Set up working directory and a virtualenv

```
mkdir vocabs
cd vocabs/
```

If you are at the Met Office do:

```
/usr/local/sci/bin/virtualenv venv
source venv/bin/activate
pip install --upgrade setuptools
pip install --upgrade pip
```

At CEDA/JASMIN, do:

```
virtualenv venv
source venv/bin/activate
```

### 2. Clone this repository and install the pyessv tool

```
git clone https://github.com/ukcp-data/UKCP18_CVs
pip install "pyessv==0.4.5.0"
```

### 3. Clone the "writer" repository 

The `pyessv-writers` repository includes a script to cache the JSON vocabularies 
in the `UKCP18_CVs` repository on to the local file system 
(under `~/.esdoc/pyessv-archive/`). Clone it and run the script:

```
git clone https://github.com/agstephens/pyessv-writers
cd pyessv-writers/
mkdir -p ~/.esdoc/pyessv-archive
python src/write_ukcp18_cvs.py --source=../UKCP18_CVs
```

### 4. Test that it works - using the pyessv library

```
$ python
Python 2.7.3 (default, Feb 21 2014, 13:11:38)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-3)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyessv
2017-07-13T07:41:27.655177 [INFO] :: ES-DOC PYESSV :: Loading vocabularies from /home/users/astephen/.esdoc/pyessv-archive:
/home/users/whoami/.esdoc/pyessv-archive/ukcp

2017-07-13T07:41:35.733418 [INFO] :: ES-DOC PYESSV :: ... loaded: ukcp
>>> import pyessv
>>> vocabs = pyessv.load('ukcp:ukcp18')
>>> insts = vocabs['institution-id']
>>> for i in insts: print i
...
ukcp:ukcp18:institution-id:ceda
ukcp:ukcp18:institution-id:mohc
ukcp:ukcp18:institution-id:stfc

```

## 5. Future usage (after the initial install)

Once you have run the above code, you only need to do the following to set up 
your environment each time you want to work with the vocabularies:

```
cd vocabs/
source venv/bin/activate
```

(And of course you could put that into a start-up file such as `.profile` or `.bashrc`).

## PYESSV usage examples

See examples at:

 https://github.com/ES-DOC/pyessv/blob/master/notebooks/pyessv-and-cmip6.ipynb

## Doing the entire install in a script

To do the above in one go, put this in a script (such as `install-vocabs.sh`), give it execute permission, and then run it.

Before you run it, set this environment variable for your `virtualenv` path:

```
# At CEDA/JASMIN:
export VIRTUALENV_DIR=/usr/bin

# At the Met Office:
export VIRTUALENV_DIR=/usr/local/sci/bin
```

Then run this script:

```
mkdir vocabs
cd vocabs/
${VIRTUALENV_DIR}/virtualenv venv
source venv/bin/activate
pip install --upgrade setuptools
pip install --upgrade pip

git clone https://github.com/ukcp-data/UKCP18_CVs
pip install pyessv

git clone https://github.com/agstephens/pyessv-writers
cd pyessv-writers/
mkdir  -p ~/.esdoc/pyessv-archive
python src/write_ukcp18_cvs.py --source=../UKCP18_CVs
```

Test it with:

```
cd vocabs/
source venv/bin/activate
python

>>> import pyessv
>>> vocabs = pyessv.load('ukcp:ukcp18')
>>> insts = vocabs['institution-id']
>>> for i in insts: print i
```


RELEASE_NOTES.md
# Release 1.0.10

This release includes:
* The addition of 'standard_name' and 'levels' to a number of variables


# Release 1.0.9

This release includes:
* Add ensemble short names


# Release 1.0.7

This release includes:
* Add collection "Regional simulations, global warming level"
* Add scenario "Global warming level of 3°C"


# Release 1.0.6

This release includes:
* Updates to humidity plot labels


# Release 1.0.5

This release includes:
* New variables jetlat, jetstr, naodjf, pr1day, pr5day


# Release 1.0.4

This release includes:
* New variables prsn, prsnAnom, snw, snwAnom


# Release 1.0.3

This release includes:
* Update to the river basins, 'Orkney and Shetlands' -> 'Orkney and Shetland'


# Release 1.0.2

This release includes:
* Update to ensemble group CVs


# Release 1.0.1

This release includes:
* Update to frequencies CVs



src/test_parse_all.py
import json, glob


def test_parsing_all_jsons():

    jsons = glob.glob("*.json")
    if len(jsons) < 5:
        jsons = glob.glob("../*.json")

    for _json in jsons:
        with open(_json) as reader:
            try:
                data = json.load(reader)
            except:
                print("FAILED TO PARSE: {0}".format(_json))


if __name__ == "__main__":

    test_parsing_all_jsons()


scripts/test_parse_all.py
import json, glob


def test_parsing_all_jsons():

    jsons = glob.glob("*.json")
    if len(jsons) < 5:
        jsons = glob.glob("../UKCP18_CVs/*.json")

    for _json in jsons:
        print('Testing: {0}'.format(_json))
        with open(_json) as reader:
            try:
                data = json.load(reader)
            except:
                print("FAILED TO PARSE: {0}".format(_json))


if __name__ == "__main__":

    test_parsing_all_jsons()


ukcp_cv/_cv.py
import json
from os import path


def enum(**named_values):
    return type('Enum', (), named_values)


CV_Type = enum(ACTIVITY_ID='activity_id',
               ADMIN_REGION='admin_region',
               BASELINE_PERIOD='baseline_period',
               CLIMATE_CHANGE_TYPE='climate_change_type',
               COLLECTION='collection',
               COORDINATE='coordinate',
               COUNTRY='country',
               DOMAIN='domain',
               ENSEMBLE_GROUP='ensemble_group',
               ENSEMBLE_MEMBER='ensemble_member',
               ENSEMBLE_SHORT_NAME='ensemble_short_name',
               FREQUENCY='frequency',
               INSTITUTION_ID='institution_id',
               LICENSE='license',
               MARINE_INPUT_MODEL='marine_input_model',
               PROB_DATA_TYPE='prob_data_type',
               PROJECT='project',
               PROJECTION='projection',
               RESOLUTION='resolution',
               RIVER_BASIN='river_basin',
               SCENARIO='scenario',
               TIME_SLICE_20y='time_slice_20y',
               TIME_SLICE_30y='time_slice_30y',
               TIME_SLICE_TYPE='time_slice_type',
               VARIABLE='variable',
               )


def get_cv(cv_type):
    """
    Get a json representation of the CV.

    @param cv_type(CV_Type): the CV of interest

    @return a json object containing details of the CV
    """
    if not _is_cv_type(cv_type):
        raise ValueError('Invalid CV_Type: {}'.format(cv_type))

    base_path = path.abspath(path.dirname(__file__))
    file_name = '../UKCP18_CVs/UKCP18_{}.json'.format(cv_type)

    cv_file = path.join(base_path, file_name)

    with open(cv_file) as json_data:
        cvs = json.load(json_data)

    return cvs


def _is_cv_type(cv_type):
    for key in CV_Type.__dict__.keys():
        if '__' not in key:
            if cv_type == CV_Type.__dict__[key]:
                return True
    return False


if __name__ == '__main__':
    print(get_cv(CV_Type.ADMIN_REGION))

    for key in CV_Type.__dict__.keys():
        if '__' not in key:
            print(key)
            print(get_cv(CV_Type.__dict__[key]))

    try:
        print(get_cv('junk'))
    except ValueError:
        pass


src/write_misc_cvs.py
#!/usr/bin/env python

"""
write_misc_cvs.py
================

Writes JSON Controlled Vocabularies for various UKCP18 terms.

"""

# Third-party imports
import simplejson
from datetime import datetime

# Local variables
now = str(datetime.now())
print now

json_tmpl = {
    "version_metadata":{
        "author":"Ag Stephens <ag.stephens@stfc.ac.uk>",
        "creation_date":now,
        "institution_id":"STFC",
        "previous_commit":""
    }
}


def _serialise(name, content):
    output_path = "UKCP18_{}.json".format(name)
    resp = json_tmpl.copy()
    resp[name] = content

    with open(output_path, 'w') as writer:
        simplejson.dump(resp, writer, indent=4, sort_keys=True)

    print "Wrote: {}".format(output_path)


def write_ensemble_members():
    data = ['r001i1p00000', 'r001i1p00090', 'r001i1p00605', 'r001i1p00834', 
            'r001i1p01113', 'r001i1p01554', 'r001i1p01649', 'r001i1p01843', 
            'r001i1p01935', 'r001i1p02089', 'r001i1p02123', 'r001i1p02242', 
            'r001i1p02305', 'r001i1p02335', 'r001i1p02491', 'r001i1p02753', 
            'r001i1p02832', 'r001i1p02868', 'r001i1p02884', 'r001i1p02914']

    dct = {}
    name = "ensemble_member"

    for key in data:
        dct[key] = key

    _serialise(name, dct)


def main():
    for func in (write_ensemble_members,):
        func()


if __name__ == "__main__":

    main()



src/climate_grid_vocabs/write_obs_cvs.py
#!/usr/bin/env python

"""
write_obs_cvs.py
================

Writes JSON CV files from the `config_tables.xls` vocabs from Climate-Grid package.

"""

# Third-party imports
import pandas
import simplejson
import numpy

CONFIG_XLS = 'config_tables.xls'
json_tmpl = {
    "version_metadata":{
        "author":"Ag Stephens <ag.stephens@stfc.ac.uk>",
        "creation_date":"Tue Jun 06 07:45:06 2017 -0100",
        "institution_id":"STFC",
        "previous_commit":""
    }
}

sheet_map = {'CF_metadata': ('variables', 'short_name'),
             'grid_properties': ('grid_properties', 'grid_name'),
             'projected_crs': ('projected_crs', 'epsg'), 
             'geographic_crs': ('geographic_crs', 'epsg')}


def _serialise(name, content):
    output_path = "../../UKCP18_obs_{}.json".format(name)
    resp = json_tmpl.copy()
    resp[name] = content

    with open(output_path, 'w') as writer:
        simplejson.dump(resp, writer, indent=4, sort_keys=True)

    print "Wrote: {}".format(output_path)


def write_json(sheet, name, key_name):
    df = pandas.read_excel(CONFIG_XLS, sheetname=sheet, na_values="")
    t = df.T # transposed
    variables = {}

    for i in range(t.shape[1]):
        data = dict(t[i])
        d = {}
    
        for key, value in data.items():
            if pandas.isnull(value): value = ""
            d[key] = value

        sn = d[key_name]

        if sn:
            variables[sn] = d.copy()

    _serialise(name, variables)


def main():
    for sheet, (name, key_name) in sheet_map.items():
        write_json(sheet, name, key_name)

if __name__ == "__main__":

    main()



scripts/write_variables_from_google_sheet.py
from __future__ import print_function

import argparse
import io
import json
import os

from apiclient import discovery
import httplib2
from oauth2client import client
from oauth2client import tools
from oauth2client.file import Storage


flags = argparse.ArgumentParser(parents=[tools.argparser]).parse_args()


# If modifying these scopes, delete your previously saved credentials
# at ~/.credentials/sheets.googleapis.com-ukcp18.json
SCOPES = 'https://www.googleapis.com/auth/spreadsheets.readonly'
CLIENT_SECRET_FILE = '../client_secret.json'
APPLICATION_NAME = 'Write UKCP18 variables from google sheets'


def get_credentials():
    """Gets valid user credentials from storage.

    If nothing has been stored, or if the stored credentials are invalid,
    the OAuth2 flow is completed to obtain the new credentials.

    Returns:
        Credentials, the obtained credential.
    """
    home_dir = os.path.expanduser('~')
    credential_dir = os.path.join(home_dir, '.credentials')
    if not os.path.exists(credential_dir):
        os.makedirs(credential_dir)
    credential_path = os.path.join(credential_dir,
                                   'sheets.googleapis.com-ukcp18.json')

    store = Storage(credential_path)
    credentials = store.get()
    if not credentials or credentials.invalid:
        flow = client.flow_from_clientsecrets(CLIENT_SECRET_FILE, SCOPES)
        flow.user_agent = APPLICATION_NAME
        credentials = tools.run_flow(flow, store, flags)
        print('Storing credentials to ' + credential_path)
    return credentials


def get_spreadsheet():
    """
    Creates a Sheets API service object and gets the contents of the
    spreadsheet.
    """
    credentials = get_credentials()
    http = credentials.authorize(httplib2.Http())
    discoveryUrl = ('https://sheets.googleapis.com/$discovery/rest?'
                    'version=v4')
    service = discovery.build('sheets', 'v4', http=http,
                              discoveryServiceUrl=discoveryUrl)
    spreadsheetId = '1Ij3R3skvYhKnMSqXB6KHaxH0BSST5R0DI8zp2Qi82vw'
    rangeName = 'Climate_variables!A2:X'
    result = service.spreadsheets().values().get(
        spreadsheetId=spreadsheetId, range=rangeName).execute()
    return result.get('values', [])


def make_dict_from_sheet(spreadsheet):
    """
    Group rows by variable id.

    @return a dict where
                key = variable id
                value = list of matching rows
    """
    id_column = 0
    data = {}
    for row in spreadsheet:
        if len(row) < 2:
            continue

        key = row[id_column].strip()
        if key in data.keys():
            data[key].append(row)
        else:
            data[key] = [row]
    return data


def process_data(data):
    """
    Create a data structure to represent a variable

    @return a dict
    """
    variable = {}
    time_step = set()
    time_averaging = set()
    notes = set()
    cmip6_cmor_tables_row_id = set()

    col_anomaly_type = 2
    col_time_step = 3
    col_long_name = 4
    col_description = 5
    col_plot_label = 6
    col_standard_name = 7
    col_units = 8
    col_label_units = 9
    col_level = 10
    col_cmip6_var_id = 11
    col_time_averaging = 12
    col_observations = 13
    col_marine = 14
    col_land_strand_1 = 15
    col_land_strand_2 = 16
    col_land_strand_3_12km = 17
    col_land_strand_3_2km = 18
    col_notes = 19
    col_um_stash = 20
    col_cmip6_standard_name = 21
    col_cmip6_cmor_tables_row_id = 22

    for row in data:
        row = [item.strip() for item in row]

        if variable == {}:
            # these fields are common to all the rows for a variable
            variable['long_name'] = row[col_long_name]
            variable['description'] = row[col_description]
            variable['plot_label'] = row[col_plot_label]

            if (row[col_anomaly_type] != "" and
                    row[col_anomaly_type] != "none"):
                variable['anomaly_type'] = row[col_anomaly_type]

            try:
                if (row[col_standard_name] != "" and
                        row[col_standard_name] != "None"):
                    variable['standard_name'] = row[col_standard_name]
                if row[col_units] != "":
                    variable['units'] = row[col_units]
                if row[col_label_units] != "":
                    variable['label_units'] = row[col_label_units]
                if row[col_level] != "":
                    variable['level'] = row[col_level]
                if (row[col_cmip6_var_id] != "" and
                        row[col_cmip6_var_id] != "None"):
                    variable['cmip6_name'] = row[col_cmip6_var_id]
                if row[col_um_stash] != "":
                    variable['um_stash'] = row[col_um_stash]
                if (row[col_cmip6_standard_name] != "" and
                        row[col_cmip6_standard_name] != "None"):
                    variable['cmip6_standard_name'] = row[col_cmip6_standard_name]
            except IndexError:
                pass

        # now sort out the differences between rows
        if row[col_time_step] != "":
            time_step.add(row[col_time_step])

        try:
            if row[col_time_averaging] != "":
                time_averaging.add(row[col_time_averaging])

            if row[col_notes] != "":
                notes.add(row[col_notes])

            if (row[col_cmip6_cmor_tables_row_id] != "" and
                    row[col_cmip6_cmor_tables_row_id] != "None"):
                cmip6_cmor_tables_row_id.add(row[col_cmip6_cmor_tables_row_id])
        except IndexError:
            pass

    if len(time_step) > 0:
        variable['time_step'] = sorted(list(time_step))
    if len(time_averaging) > 0:
        variable['time_averaging'] = sorted(list(time_averaging))
    if len(notes) > 0:
        variable['notes'] = sorted(list(notes))
    if len(cmip6_cmor_tables_row_id) > 0:
        variable['cmip6_cmor_tables_row_id'] = sorted(
            list(cmip6_cmor_tables_row_id))

    return variable


def main():
    spreadsheet = get_spreadsheet()
    variables = {}
    data = make_dict_from_sheet(spreadsheet)
    for key in data.keys():
        variables[key] = process_data(data[key])

    _add_special_variables(variables)

    output = {'variable': variables}

    variable_json = '../UKCP18_CVs/UKCP18_variable.json'
    with io.open(variable_json, 'w', encoding='utf8') as json_file:
        data = json.dumps(output, sort_keys=True, indent=4,
                          separators=(',', ': '), ensure_ascii=False)
        json_file.write(data)

    print('Wrote new version of: {}'.format(variable_json))


def _add_special_variables(variables):
    """
    Adds in special variables defined with non-standard attributes.

    :param variables: dictionary.
    :return: None
    """
    """
        int transverse_mercator ;
grid_mapping_name = "transverse_mercator" ;
longitude_of_prime_meridian = 0. ;
semi_major_axis = 6377563.396 ;
semi_minor_axis = 6356256.909 ;
longitude_of_central_meridian = -2. ;
latitude_of_projection_origin = 49. ;
false_easting = 400000. ;
false_northing = -100000. ;
scale_factor_at_central_meridian = 0.9996012717 ;
    """
    tv = {"grid_mapping_name": "transverse_mercator",
          "longitude_of_prime_meridian": 0.,
          "semi_major_axis": 6377563.396,
          "semi_minor_axis": 6356256.909,
          "longitude_of_central_meridian": -2.,
          "latitude_of_projection_origin": 49.,
          "false_easting": 400000.,
          "false_northing": -100000.,
          "scale_factor_at_central_meridian": 0.9996012717,
          }
    variables['transverse_mercator'] = tv


if __name__ == '__main__':
    main()
